pallets/funding/src/settlement.rs

```diff
@@ -0,0 +1,306 @@
+use super::*;
+
+use frame_support::{
+	dispatch::{ DispatchResult},
+	ensure,
+	pallet_prelude::*,
+	traits::{
+		fungible::{InspectHold, MutateHold as FungibleMutateHold},
+		fungibles::{
+            Inspect, Mutate as FungiblesMutate,
+		},
+		tokens::{Fortitude, Precision, Preservation, Restriction},
+		Get,
+	},
+};
+use sp_runtime::{traits::Zero, Perquintill};
+use polimec_common::{
+	migration_types::{MigrationInfo, MigrationOrigin, Migrations, ParticipationType},
+	ReleaseSchedule,
+};
+
+impl<T: Config> Pallet<T> {
+    pub fn do_settlement_success_bidder(bid: BidInfoOf<T>, project_id: ProjectId) -> DispatchResult {
+        
+        let project_details = ProjectsDetails::<T>::get(project_id).ok_or(Error::<T>::ProjectDetailsNotFound)?;
+		// Ensure that:
+        // 1. The project is in the FundingSuccessful state
+        // 2. The bid is in the Accepted or PartiallyAccepted state
+        // 3. The contribution token exists
+        ensure!(project_details.status == ProjectStatus::FundingSuccessful, Error::<T>::NotAllowed);
+        ensure!(matches!(bid.status, BidStatus::Accepted | BidStatus::PartiallyAccepted(..)), Error::<T>::NotAllowed);
+        ensure!(T::ContributionTokenCurrency::asset_exists(project_id), Error::<T>::CannotClaimYet);
+
+        let bidder = bid.bidder;
+
+        // Calculate the vesting info and add the release schedule
+        let funding_end_block = project_details.funding_end_block.ok_or(Error::<T>::ImpossibleState)?;
+		let vest_info =
+			Self::calculate_vesting_info(&bidder, bid.multiplier, bid.plmc_bond).map_err(|_| Error::<T>::BadMath)?;
+		
+        T::Vesting::add_release_schedule(
+            &bidder,
+            vest_info.total_amount,
+            vest_info.amount_per_block,
+            funding_end_block,
+            HoldReason::Participation(project_id).into(),
+        )?;
+
+        // Mint the contribution tokens
+        Self::mint_ct_tokens(project_id, &bidder, bid.final_ct_amount)?;
```

Comment:
are we not using the functions associated with the extrinsics? like `do_bid_ct_mint_for`?

Reply:
All these functions had a lot of overlapping code so I changed nearly 800 lines of function code to around 300.